include/nbl/video/IGPUDescriptorSet.h

```diff
@@ -28,10 +30,131 @@ class NBL_API IGPUDescriptorSet : public asset::IDescriptorSet<const IGPUDescrip
+            if (baseAddress == nullptr)
+                return nullptr;
+
+            const auto poolOffset = m_descriptorStorageOffsets.data[asset::EDT_COUNT];
+            if (poolOffset == ~0u)
+                return nullptr;
+
+            return baseAddress + poolOffset;
+        }
+
+        // This assumes that descriptors of a particular type in the set will always be contiguous in pool's storage memory, regardless of which binding in the set they belong to.
+        inline core::smart_refctd_ptr<asset::IDescriptor>* getDescriptors(const asset::E_DESCRIPTOR_TYPE type, const uint32_t binding) const
+        {
+            const auto localOffset = getLayout()->getDescriptorOffset(type, binding);
+            if (localOffset == ~0)
+                return nullptr;
+
+            auto* descriptors = getAllDescriptors(type);
+            if (!descriptors)
+                return nullptr;
+
+            return descriptors + localOffset;
+        }
+
+        inline core::smart_refctd_ptr<IGPUSampler>* getMutableSamplers(const uint32_t binding) const
+        {
+            const auto localOffset = getLayout()->getMutableSamplerOffset(binding);
+            if (localOffset == ~0u)
+                return nullptr;
+
+            auto* samplers = getAllMutableSamplers();
+            if (!samplers)
+                return nullptr;
+
+            return samplers + localOffset;
+        }
+
+        inline uint32_t getDescriptorStorageOffset(const asset::E_DESCRIPTOR_TYPE type) const { return m_descriptorStorageOffsets.data[type]; }
+        inline uint32_t getMutableSamplerStorageOffset() const { return m_descriptorStorageOffsets.data[asset::EDT_COUNT]; }
 
 	protected:
-		virtual ~IGPUDescriptorSet() = default;
+		virtual ~IGPUDescriptorSet()
+		{
+            for (auto i = 0u; i < asset::EDT_COUNT; ++i)
+            {
+                // There is no descriptor of such type in the set.
+                if (m_descriptorStorageOffsets.data[i] == ~0u)
+                    continue;
+
+                const auto type = static_cast<asset::E_DESCRIPTOR_TYPE>(i);
+                std::destroy_n(getDescriptorStorage(type) + m_descriptorStorageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
+            }
+
+            const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
+            if (mutableSamplerCount > 0)
+                std::destroy_n(getMutableSamplerStorage() + getMutableSamplerStorageOffset(), mutableSamplerCount);
+		}
+
+	private:
+        struct SDescriptorOffsets
+        {
+            SDescriptorOffsets()
+            {
+                // The default constructor should initiailze all the offsets to an invalid value (~0u) because ~IGPUDescriptorSet relies on it to
+                // know which descriptors are present in the set and hence should be destroyed.
+                std::fill_n(data, asset::EDT_COUNT + 1, ~0u);
+            }
+
+            uint32_t data[asset::EDT_COUNT + 1];
+        };
```

Comment:
probably need to move to `IDescriptorPool` so it can give them out

Reply:
Yes, resolved.